app/lib/Data.scala

```diff
@@ -0,0 +1,128 @@
+package lib
+
+import scala.util.{Failure, Success, Try}
+import java.math.BigInteger
+
+import play.api.mvc._
+import play.api.libs.json._
+import play.api.mvc.Results._
+import com.github.nscala_time.time.Imports._
+import reactivemongo.bson.BSONObjectID
+
+
+
+object Data {
+
+  // Data type definitions ------------------------------------------------------------------------
+  type ScreeningId  = BSONObjectID
+  type RoomId       = String
+  type RowId        = Int
+  type ColumnId     = Int
+
+
+  case class Seat(row: RowId, column: ColumnId)
+
+
+  sealed trait Error
+  final object InvalidParameters      extends Error
+  final object SeatsAlreadyTaken      extends Error
+  final object SeatsNotConnected      extends Error
+  final object InconsistentData       extends Error
+  final object NoSuchScreening        extends Error
+  final object InvalidBody            extends Error
+  final object InvalidScreeningId     extends Error
+  final object Unknown                extends Error
+  final object TooLateForReservation  extends Error
+  final object InvalidName            extends Error
+  final object InvalidSurname         extends Error
+  final object SeatOutOfRange         extends Error
+  final object EmptyReservation       extends Error
+
+
+  sealed trait TicketType
+  final object Adult    extends TicketType
+  final object Student  extends TicketType
+  final object Child    extends TicketType
+
+  final case class RoomDimension(numRows: Int, numColumns: Int)
+  
+  final case class AvailableSeats(dim: RoomDimension, seats: List[List[ColumnId]])
+  final case class TakenSeats(dim: RoomDimension, seats: List[Seat])
+
+  final case class Person(name: String, surname: String)
+
+  final case class ScreeningInfo(
+    id:       ScreeningId,
+    title:    String,
+    start:    DateTime,
+    duration: Duration,
+    room:     RoomId,
+  )
+
+  final case class Reservation(
+    screening:  ScreeningId,
+    seats:      Map[Seat, TicketType],
+    reserver:   Person,
+  )
+
+  // Utils ----------------------------------------------------------------------------------------
+  def ticketPrice(ticket: TicketType): Double = ticket match {
+    case Adult    => 25
+    case Student  => 18
+    case Child    => 12.5
+  }
```

Comment:
Do you know any way to pass this information but not hardcode it?

Reply:
I could put it in a config file. In real life the prices would probably be stored in the database.